packages/fast-components-react-base/src/tabs/tabs.tsx

```diff
@@ -0,0 +1,336 @@
+        const nextItemId: string = items[nextItemIndex].props.id;
+
+        if (!this.props.activeId) {
+            this.setState({
+                activeId: nextItemId
+            });
+
+            (Array.from(this.tabListRef.current.children)[nextItemIndex] as HTMLButtonElement).focus();
+        } else if (typeof this.props.onUpdateTab === "function") {
+            this.props.onUpdateTab(nextItemId);
+        }
+    }
+
+    /**
+     * Activates the first tab item
+     */
+    private activateFirst(): void {
+        const items: JSX.Element[] = this.getChildBySlot(
+            this.props.children,
+            this.getSlot(TabSlot.tabItem)
+        );
+        const activeId: string = items[0].props.id;
+
+        if (!this.props.activeId) {
+            this.setState({
+                activeId
+            });
+
+            (Array.from(this.tabListRef.current.children)[0] as HTMLButtonElement).focus();
+        } else if (typeof this.props.onUpdateTab === "function") {
+            this.props.onUpdateTab(activeId);
+        }
+    }
+
+    /**
+     * Activates the last tab item
+     */
+    private activateLast(): void {
+        const items: JSX.Element[] = this.getChildBySlot(
+            this.props.children,
+            this.getSlot(TabSlot.tabItem)
+        );
+        const lastItemIndex: number = items.length - 1;
+        const activeId: string = items[lastItemIndex].props.id;
+
+        if (!this.props.activeId) {
+            this.setState({
+                activeId
+            });
+
+            (Array.from(this.tabListRef.current.children)[lastItemIndex] as HTMLButtonElement).focus();
+        } else if (typeof this.props.onUpdateTab === "function") {
+            this.props.onUpdateTab(activeId);
+        }
+    }
+
+    /**
+     * Gets the current index by tab item ID
+     */
+    private getCurrentIndexById = (item: JSX.Element): boolean => {
+        return item.props.id === this.state.activeId;
+    }
+
+    /**
+     * Gets the slot to use for tab children
+     */
+    private getSlot(slot: TabSlot): TabSlot | string {
+        switch (slot) {
+            case TabSlot.tab:
+                return typeof this.props.tabSlot === "string" ? this.props.tabSlot : TabSlot.tab;
+            case TabSlot.tabItem:
+                return typeof this.props.tabItemSlot === "string" ? this.props.tabItemSlot : TabSlot.tabItem;
+            case TabSlot.tabPanel:
+                return typeof this.props.tabPanelSlot === "string" ? this.props.tabPanelSlot : TabSlot.tabPanel;
+        }
+    }
+
+    /**
+     * Gets the child by the slot property
+     */
+    private getChildBySlot(children: React.ReactNode, slot: TabSlot | string): JSX.Element[] {
+        const childBySlot: JSX.Element[] = [];
+
+        React.Children.forEach(children, (child: JSX.Element): void => {
+            if (child.props && child.props.slot === slot) {
+                if (slot === this.getSlot(TabSlot.tabItem)) {
+                    if (
```

Comment:
I'd also reccomend moving the bit of logic to filter out `TabItem` elements that don't have both a `tabPanel` and `tab` to another function. The name of this function implies that I'm just returning children that match a particular slot.

That keeps this function single-purpose and provides a semantic name to tie to the behavior of only rendering components that have both `tab` and `tabPanel` elements.